### PR TITLE
fuse qkvbfg linear into one gemm and f_b g_b into batched gemm.

### PR DESCRIPTION
**Pull request by strgrb**
## Motivation

There are 8 gemm in kda, and 6 of them (q_proj, k_proj, v_proj, b_proj, f_a_proj, g_a_proj) share the same input, so they can be fused into single gemm, and other 2 of them (f_b_proj, g_b_proj) has the same input/output size, so they can be fused into a batched gemm.

With 4k input 1k output test for decode, the profile like the following:
image

kernel duration from 5.9+5.7+5.6+5.6+2.5+5.2+5.7+2.6=38.8us to 8.7+3=11.7us

For prefill, kernel duration from 286+286+286+45+20+14+45+19=1001us to 934+35=969us, optimization is much less than decode.

## Accuracy Tests

test with gsm8k all questions:
```
python bench_sglang.py --port 8189 --data-path ./test.jsonl --num-questions -1
```

before:
```
Accuracy: 0.895
Invalid: 0.001
Latency: 109.134 s
Output throughput: 1175.404 token/s
```
after:
```
Accuracy: 0.898
Invalid: 0.000
Latency: 104.709 s
Output throughput: 1224.809 token/s
```

## Benchmarking and Profiling

before:
max concurrency | random input len | random output len | throughput | TTFT | TPOT 
----|----|----|----|----|----
8   | 4000 | 1000 | 1.26 | 473.58 | 5.88
32   | 4000 | 1000 | 2.50 | 1445.09 | 11.33
128   | 4000 | 1000 | 4.38 | 5326.57 | 23.62

after:
max concurrency | random input len | random output len | throughput | TTFT | TPOT 
----|----|----|----|----|----
8   | 4000 | 1000 | 1.37 | 470.89 | 5.28
32   | 4000 | 1000 | 2.65 | 1443.90 | 10.62
128   | 4000 | 1000 | 4.56 | 5612.32 | 22.48

TPOT is decreased by about 5%~10%

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/kimi_linear.py [hunk 1]
```diff
@@ -160,6 +163,89 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         return final_hidden_states.view(num_tokens, hidden_size)
 
 
+class MergedColumnParallelRepeatedLinear(LinearBase):
```
Comment (yuan-luo) [inline python/sglang/srt/models/kimi_linear.py]: Maybe put it into srt/layers/linear.py to benefit more modules?

File: python/sglang/srt/models/kimi_linear.py [hunk 2]
```diff
@@ -160,6 +163,89 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
+        param_data = param.data.narrow(output_dim, shard_offset, shard_size)
+
+        if loaded_shard_id < self.num_column_parallel:
+            start_idx = self.tp_rank * shard_size
+            loaded_weight = loaded_weight.narrow(output_dim, start_idx, shard_size)
+
+        param_data.copy_(loaded_weight)
+
+
+class ColumnParallelBatchedLinear(nn.Module):
```
Comment (yuan-luo) [inline python/sglang/srt/models/kimi_linear.py]: Likewise.

File: python/sglang/srt/models/kimi_linear.py [hunk 3]
```diff
@@ -190,57 +192,97 @@ def __init__(
...
+        self.do_fuse_qkvbfg = quant_config is None
+        if self.do_fuse_qkvbfg:
+            self.qkvb_sizes = [
+                projection_size,
+                projection_size,
+                projection_size,
+                self.num_heads,
+            ]
+            self.fg_sizes = [self.head_dim, self.head_dim]
+            self.fused_qkvbfg_proj = MergedColumnParallelRepeatedLinear(
```
Comment (ispobock) [inline python/sglang/srt/models/kimi_linear.py:205]: It seems that, for large bs, it will cause higher ttft, since it's already compute bound.

Comment (strgrb) [reply python/sglang/srt/models/kimi_linear.py:205]: Gemm with large M will not cause higher latency, I think this benchmark result may be caused by random issue. I'll test it for more times and do some separate test.

Comment (strgrb) [reply python/sglang/srt/models/kimi_linear.py:205]: @ispobock I used benchmark but ttft result is not so stable, so I use throughput at output_size=1 to identify the prefill performance at batch_size=128.
Before opt is 12.87 and after is 12.96, this result is quite stable.

Comment (strgrb) [reply python/sglang/srt/models/kimi_linear.py:205]: @ispobock And for separate test, I use following code to test. Mainly for m=4096 to 16384, because these values are commonly used `chunked_prefill_size` .
```
import torch

m = 8192
k = 2304
n1s = [1024, 1024, 1024, 8, 128, 128]
n2 = 1024

x = torch.randn([m, k], device='cuda', dtype=torch.bfloat16)
w1s = [torch.randn([n, k], device='cuda', dtype=torch.bfloat16) for n in n1s]
w2s = [torch.randn([n2, 128], device='cuda', dtype=torch.bfloat16) for _ in range(2)]

merged_w1 = torch.cat(w1s, dim=0)
merged_w2 = torch.stack(w2s, dim=0)

def forward_before():
  y1s = [x @ w.T for w in w1s]
  y21 = y1s[-2] @ w2s[0].T
  y22 = y1s[-1] @ w2s[1].T

def forward_after():
  merged_y1 = x @ merged_w1.T
  merged_y2 = torch.bmm(merged_y1[:, -256:].view(m, 2, 128).transpose(0, 1), merged_w2.transpose(-1, -2))

forward_before()
forward_after()
```

For m=4096, 0.57ms vs 0.50ms, for m=8192, 1.03ms vs 0.98ms, for m=16384, 1.99ms vs 1.91ms

**Review by ispobock — APPROVED** [review]
